text/3735-field-projections.md

```diff
@@ -0,0 +1,1344 @@
+
+let _ = Project::<field_of!(<C<T> as Projectable>::Inner, field)>::project(t);
+```
+
+The `C<T>` in the `C<T> as Projectable` comes from a type inference variable over the expression
+`t`.
+
+
+#### `Projectable` & `Project` Traits
+
+Added to `core::ops`:
+
+```rust
+pub trait Projectable: Sized {
+    type Inner: ?Sized;
+}
+
+pub trait Project<F>: Projectable
+where
+    F: Field<Base = Self::Inner>,
+{
+    type Output;
+
+    fn project(self) -> Self::Output;
+}
+```
+
+## Stdlib Field Projections
+
+All examples from the guide-level explanation work when the standard library is extended with the
+implementations detailed below.
+
+The following pointer types get an implementation for `Projectable` with `Inner = T`. They support
+projections for any field and perform the obvious offset operation.
+
+- `&mut T`
+- `&T`
+- `*mut T`
+- `*const T`
+- `NonNull<T>`
+- `Cow<'_, T>`
+
+The following types get annotated with [`#[projecting]`](#projecting-attribute):
+
+- `MaybeUninit<T>`
+- `Cell<T>`
+- `UnsafeCell<T>`
+- `SyncUnsafeCell<T>`
+
+### Pin Projections
+
+In order to provide [pin projections], a new derive macro `PinProject` and a trait `PinField` is
+required:
+
+```rust
+pub unsafe trait PinField: Field {
+    type Projected<'a>;
+
+    fn from_pinned_ref<'a>(r: &'a mut Self::Type) -> Self::Projected<'a>;
+}
+```
+
+The trait is `unsafe`, because it must only be implemented by the `PinProject` derive marco.
+
+```rust
+#[derive(PinProject)]
+struct FairRaceFuture<F1, F2> {
+    #[pin]
+    fut1: F1,
+    #[pin]
+    fut2: F2,
+    fair: bool,
+}
+```
+
+It expands the above to:
+
+```rust
+struct FairRaceFuture<F1, F2> {
+    fut1: F1,
+    fut2: F2,
+    fair: bool,
+}
+
+unsafe impl<F1, F2> PinField for field_of!(FairRaceFuture<F1, F2>, fut1) {
+    type Projected<'a> = Pin<&'a mut F1>;
+
+    fn from_pinned_ref<'a>(r: &'a mut F1) -> Pin<&'a mut F1> {
+      unsafe { Pin::new_unchecked(r) }
+    }
+}
+
+unsafe impl<F1, F2> PinField for field_of!(FairRaceFuture<F1, F2>, fut2) {
+    type Projected<'a> = Pin<&'a mut F2>;
+
+    fn from_pinned_ref<'a>(r: &'a mut F2) -> Pin<&'a mut F2> {
+      unsafe { Pin::new_unchecked(r) }
+    }
+}
+
+unsafe impl<F1, F2> PinField for field_of!(FairRaceFuture<F1, F2>, fut2) {
+    type Projected<'a> = &'a mut bool;
+
+    fn from_pinned_ref<'a>(r: &'a mut bool) -> &'a mut bool {
+      r
+    }
+}
+```
+
+Now the only component that is left is an implementation of `Projectable` and `Project` for
+`Pin<&mut T>`:
+
+```rust
+impl<'a, T: ?Sized> Projectable for Pin<&'a mut T> {
+    type Inner = T;
+}
+
+impl<'a, T, F> Project<F> for Pin<&'a mut T>
+where
+    F: Field<Base = T> + PinField,
+{
+    type Output = <F as PinField>::Projected<'a>;
+
+    fn project(self) -> Self::Output {
+        let r: &mut T = unsafe { Pin::into_inner_unchecked(self) };
+        let r: &mut F::Type = <&mut T as Project::<F>>::project(r);
+        <F as PinField>::from_pinned_ref(r)
+    }
+}
+```
+
+## Interactions
```

Comment:
One question I didn't see mentioned: what about simultaneous projections? e.g.
```rust
#[derive(PinProject)]
struct FairRaceFuture<F1, F2> {
    #[pin]
    fut1: F1,
    #[pin]
    fut2: F2,
    fair: bool,
}
let fut: Pin<&mut FairRaceFuture<_, _>> = ...;
let fut1 = fut->fut1;
let fut2 = fut->fut2;
```
Most projectable pointers would theoretically allow `fut1` and `fut2` to coexist. But that seems really hard to allow with user-provided types: `fut` should clearly not remain fully accessible while `fut1` is live, yet passing it to a function like `project` likely invalidates `fut1`. We'd need `project` to take raw pointers, but then we can't have it for custom types.

Is there a design that could accomodate this in some way?

Reply:
> One question I didn't see mentioned: what about simultaneous projections? e.g.

That is indeed something that I should mention.

> Most projectable pointers would theoretically allow `fut1` and `fut2` to coexist. But that seems really hard to allow with user-provided types: `fut` should clearly not remain fully accessible while `fut1` is live, yet passing it to a function like `project` likely invalidates `fut1`. We'd need `project` to take raw pointers, but then we can't have it for custom types.
> 
> Is there a design that could accomodate this in some way?

In the [Design Meeting on Field Projections](https://hackmd.io/@y86-dev/SkkfRkzWh#Problem-4-Automatic-Reborrow), I brought this up, but only for the `Pin` case. There one needs some kind of "automatic reborrowing". But this still only allows consecutive uses of projected fields, not simultaneous usage.

For simultaneous usage, we would have to involve the borrow checker, and let the user choose a behavior (i.e. should it be like `&T`, `&mut T`, `*mut T`. maybe there are more?). That might be a good way to fix this, either through a trait or through macro intrinsics (that depends on how the borrow checker works internally).

```rust
borrow_check!(unsafe <'a, T> Pin<&'a mut T> as &'a mut T);
// or
unsafe impl<'a, T> BorrowCheck<&'a mut T> for Pin<&'a mut T> {}
```

Reply:
Yes, that is correct, but if we do that with `&mut Struct` instead, the borrow checker knows what is going on:

```rust
let x: &mut Struct = ...;
let field1 = &mut x.field1;
let field2 = &mut x.field2;
```

We just have to make `Pin<&mut T>` behave in the same way as `&mut T` for the borrow checker.

Reply:
Agreed that we could tell the borrow-checker to consider `fut->field1` and `fut->field2` to be disjoint, but then we'd need the underlying memory model to allow this without UB. The problem we have is the `project` function call. As far as I know of the current state of discussion on this, passing a `&mut Whatever` to a function morally counts as an access, which invalidates any sibling references to that `Whatever`. I'll let the opsem experts clarify though.

This is similar to how we can't borrow two disjoint fields through a `DerefMut` (`Box` cheats on this btw).

Reply:
Ah yes, that is a problem...

Reply:
Maybe what we need is:
```rust
/// SAFETY: An impl is safe if calling `project` several times on the same `self` but disjoint fields and exposing the resulting outputs to safe code is safe.
unsafe pub trait ProjectMut<F: Field<Base = Self::Inner>>: Projectable {
    type Output;
    /// SAFETY: The input must be cast from a safe `&mut Self`, and possibly passed to this `project` function with disjoint `F` fields and nothing else.
    unsafe fn project(*mut self) -> Self::Output;
}
```
The safety requirements would need fleshing out to be rigorous but you get the idea: the borrow-checker would ensure we don't hold two projected pointers to the same field at the same time.

There'd also be a `ProjectShared` for the shared case, that allows multiple aliasing shared projections. And `ProjectMove` would ask the borrow-checker to track which fields have been moved out and to project+drop the non-moved fields at end of scope.

Reply:
> As far as I know of the current state of discussion on this, passing a &mut Whatever to a function morally counts as an access, which invalidates any sibling references to that Whatever. I'll let the opsem experts clarify though.

This is correct. We'd want this to be more like a `MaybeDangling<&mut T>`, and suppress all aliasing-related requirements temporarily, until we have the field and it can become an `&mut F`.